### PR TITLE
#176 EA-2014 implemented new monolingual algorithm for suggester

### PR DESCRIPTION
**Pull request by rgraf**

### CODE REVIEW

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java [hunk 1]
```diff
@@ -125,6 +132,33 @@ public SolrQuery buildSuggestByLabelQuery(String text, List<EntityTypes> entityT
...
+
+	/**
+	 * This method selects entities by label and provided languages, which should be included in the
+	 * response
+	 * 
+	 * @param label
+	 * @param languages
+	 * @return Solr query for label by provided languages
+	 */
+	private String buildQueryByLabelAndLanguages(String label, List<String> languages) {
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:145]: please rename to buildQueryByLabel

Comment (gsergiu) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:145]: please rename label param to text

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java [hunk 2]
```diff
@@ -125,6 +132,33 @@ public SolrQuery buildSuggestByLabelQuery(String text, List<EntityTypes> entityT
...
+		return "(" + disjunction + ")";
+	}	
+	
+	/**
+	 * This method builds a query for label and one language
+	 * @param label
+	 * @param language
+	 * @return
+	 */
+	private static String buildOneLanguageLabelQuery(String label, String language) {
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:159]: please rename to buildQuery and label param to text

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java [hunk 3]
```diff
@@ -160,6 +194,20 @@ private String buildEntityTypeCondition(List<EntityTypes> entityTypes) {
 		return internalEntityTypes;
 	}
 
+	/**
+	 * Convert a list of languages for label into an array of strings
+	 * 
+	 * @param languages The array of languages
+	 * @return array of strings
+	 */
+	static String[] toLabelsStringArray(String label, List<String> languages) {
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:203]: please remove this method and include the for loop in the calling method

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java [hunk 4]
```diff
@@ -289,6 +312,88 @@ public Logger getLogger() {
...
+     * 
+     * @param searchedTerm
+     * @param rsp
+     * @param requestedLanguages
+     * @param rows
+     * @return entity preview
+     * @throws EntitySuggestionException
+     */
+    @SuppressWarnings({ "unchecked" })
+    protected <T extends EntityPreview> ResultSet<T> buildSuggestionSetForSearchByLanguages(String searchedTerm,
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java:327]: rename to buildSuggestionSet

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java [hunk 5]
```diff
@@ -289,6 +312,88 @@ public Logger getLogger() {
...
+     * 
+     * @param searchedTerm
+     * @param rsp
+     * @param requestedLanguages
+     * @param rows
+     * @return entity preview
+     * @throws EntitySuggestionException
+     */
+    @SuppressWarnings({ "unchecked" })
+    protected <T extends EntityPreview> ResultSet<T> buildSuggestionSetForSearchByLanguage(String searchedTerm,
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java:368]: this method is a dupplicated implementation of the previous method, please remove it

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java [hunk 6]
```diff
@@ -77,6 +77,34 @@ public EntityPreview parsePayload(String payload, String[] preferredLanguages, S
...
 
+	/**
+	 * This method parses a payload employing preferred languages and highlighted terms.
+	 * @param payload
+	 * @param preferredLanguages
+	 * @param highlightTerms
+	 * @return parsed payload
+	 * @throws EntitySuggestionException
+	 */
+	public EntityPreview parsePayloadByLanguage(String payload, String[] preferredLanguages, Set<String> highlightTerms)
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java:88]: this seems to be a dupplicated implementation ofthe parsePyloadMethod, please remove

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java [hunk 7]
```diff
@@ -134,7 +162,63 @@ private EntityPreview parseEntity(JsonNode entityNode,  List<String> preferredLa
...
-	
+	/**
+	 * This method parses entity employing preferred languages and highlighted terms.
+	 * @param entityNode
+	 * @param preferredLanguages
+	 * @param highlightTerms
+	 * @return parsed entity
+	 * @throws UnsupportedEntityTypeException
+	 */
+	private EntityPreview parseEntityByLanguage(JsonNode entityNode,  List<String> preferredLanguages, Set<String> highlightTerms)
```
Comment (gsergiu) [inline entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java:173]: this is a dupplicated implementation parseEntity, please remove this method

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java (hunk 2 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:159]: renamed to buildQuery

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java (hunk 7 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java:173]: removed duplicated method

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java (hunk 6 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SuggestionUtils.java:88]: removed duplicated method

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java (hunk 4 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/SolrEntityServiceImpl.java:327]: renamed to buildSuggestionSet

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java (hunk 3 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:203]: for loop included in the calling method

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java (hunk 1 above)
Comment (rgraf) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:145]: renamed required parameters

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityConstants.java [hunk 8]
```diff
@@ -71,6 +71,7 @@
 	public static final String ALGORITHM = "algorithm";
 	//see SuggestAlgorithmTypes.suggestByLabel
 	public static final String SUGGEST_ALGORITHM_DEFAULT = "suggestByLabel"; 
+	public static final String SUGGEST_ALGORITHM_BY_LANGUAGE = "suggestByLanguage";
```
[the remark this answers is not in the document]
Comment (rgraf) [reply entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityConstants.java:74]: renamed to suggestForLanguage

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/SuggestAlgorithmTypes.java [hunk 9]
```diff
@@ -4,7 +4,7 @@
 
 public enum SuggestAlgorithmTypes {
 
-	suggest, suggestByLabel;
+	suggest, suggestByLabel, suggestByLanguage;
```
[the remark this answers is not in the document]
Comment (rgraf) [reply entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/SuggestAlgorithmTypes.java:7]: renamed to suggestForLanguage

**Pull request merged**